src/vs/workbench/contrib/search/browser/searchActions.ts

```diff
@@ -439,6 +439,139 @@ export abstract class AbstractSearchAndReplaceAction extends Action {
 
 		return previousElement;
 	}
+
+}
+class ReplaceActionRunner {
+	constructor(
+		private viewer: WorkbenchObjectTree<RenderableMatch>,
+		private viewlet: SearchView | undefined,
+		private getElementToFocusAfterRemoved: (viewer: WorkbenchObjectTree<RenderableMatch>, elementToBeRemoved: RenderableMatch) => RenderableMatch,
+		private getPreviousElementAfterRemoved: (viewer: WorkbenchObjectTree<RenderableMatch>, element: RenderableMatch) => RenderableMatch,
+		// Services
+		@IReplaceService private readonly replaceService: IReplaceService,
+		@IEditorService private readonly editorService: IEditorService,
+		@IConfigurationService private readonly configurationService: IConfigurationService,
+		@IUriIdentityService private readonly uriIdentityService: IUriIdentityService
+	) { }
+
+	async performReplace(element: FolderMatch | FileMatch | Match): Promise<any> {
+		// since multiple elements can be selected, we need to check the type of the FolderMatch/FileMatch/Match before we perform the replace.
+
+		const opInfo = getElementsToOperateOnInfo(this.viewer, element);
+
+		await Promise.all(opInfo.elementsToRemove.map(async (elem) => {
+			const parent = elem.parent();
+
+			if ((parent instanceof FolderMatch || parent instanceof FileMatch) && arrayContainsElementOrParent(parent, opInfo.elementsToRemove)) {
+				// skip any children who have parents in the array
+				return;
+			}
+
+			if (parent instanceof FolderMatch && elem instanceof FileMatch) {
+				await (<FolderMatch>(elem.parent())).replace(<FileMatch>elem);
```

Comment:
The casts shouldn't be necessary - inside the `if` with `instanceof`, TS knows what the type is because you just checked it. I think you also don't need these two if blocks that have essentially the same code (`parent.replace(child)`). I'd think you can check the replaceAll case first, then cover the rest with `else { ...`

Reply:
is this what you were envisioning? 
![image](https://user-images.githubusercontent.com/31675041/179606333-2ec89527-8089-466b-a37e-f07ab4cf7f18.png)

The replace ends up requiring a type `never`.

Reply:
I guess the model types are not designed in a nice way to allow things like this. I think what you basically have is fine, except that there should be no casts, and you only need to check the type of `elem`. Checking the type of the parent that you just retrieved from `elem` is redundant. Example:

```ts
if (elem instanceof FileMatch) {
	elem.parent().replace(elem);
} else if (elem instanceof Match) {
	elem.parent().replace(elem);
} else {
	await elem.replaceAll();
}
```

If you feel like mucking about with TS you could probably reduce the first two cases to one case.

Reply:
I believe that, if I wanted to combine the first two cases like `elem instanceof FileMatch || elem instanceof Match`, the `replace` method would take `never ` and I would need to cast to `any`. I think I'll leave mine as what you suggested above